docs/architecture.md

```diff
@@ -0,0 +1,271 @@
+the simulation. Components are a very flexible mechanism, with the
+only requirement being that a component provides an `init()` function
+which is invoked when the component is first added to the `Context`.
+
+It is possible for a component to do all of its work inside `init()`.
+For instance, if you had a component which loaded a population into
+the simulation, that component would probably do that in `init()`
+because you want the population loaded prior to the simulation
+beginning. More complex components will want to do work *during* the
+simulation not just prior to it. In order to do so, they register one
+or more `Plan`s, which do the work of the simulation.
+
+
+## Scheduling Plans
+
+A `Plan` is a callback which is executed at some deterministic point
+of time in the future. For instance, if we wanted to model COVID
+isolation guidance, when someone is infected we might set a `Plan` to
+have them take a rapid antigen test 48 hours after becoming
+symptomatic.
+
+In order to schedule a `Plan` a component provides:
+
+* The *execution time* which is when the plan will execute
+* The *callback* to fire
+
+More details about callbacks are provided in [TODO] but at a high
+level, callbacks are Rust functions (typically closures) which
+get passed a reference to the `Context` as an argument. This
+allows them to act on the `Context` or any of its attached
+objects, such as data containers provided by plugins.
+
+A `Plan`'s callback can schedule other plans, thus allowing
+the simulation to condition future events on current events.
+In general, a `Component` will just work out the next action
+for a specific subject (e.g., a person in the simulation)
+and set a `Plan` for that. Once that action has taken place,
+the `Component` will then work out what happens next; it
+is generally not desirable to work out a detailed list
+of actions for each subject well in advance, as conditions
+may change rendering that planning obsolete.
+
+It is possible to schedule two `Plan`s for the same execution
+time. In this case, Eosim will run them in the order scheduled;
+in the future we might provide mechanisms to break this kind
+of tie.
+
+### Canceling Plans
+
+Even so, in some cases, a `Component` will make a plan but then
+events will happen in the intervening time that make it
+irrelevant. For instance, in the above example where we make
+a `Plan` for a person to take a test in 48 hours, what
+happens if that person dies in the intervening time?
+One way to handle this is to have the callback check
+if the person is alive and only then take the test,
+but this is not very convenient, as it pushes logic into the callback.
+
+Instead, Eosim provides a say to *cancel* a `Plan`: when
+a `Plan` is scheduled, the `Context` provides a `PlanId`
+which can then be used to cancel the `Plan`. Note that
+this requires storing the `PlanId` somewhere, typically
+in a `DataContainer`.
+
+### Immediate Callbacks
+
```

Comment:
I might add to this section why we need these kinds of immediate callbacks. This queue is present to capture side effects of the plan that is currently being processed. The inherent assumption is that these side effects take priority of previously-scheduled plans for the same time. The canonical example is a plan that modifies a person property that some Component has indicated it would like to handle events where the property is changed.